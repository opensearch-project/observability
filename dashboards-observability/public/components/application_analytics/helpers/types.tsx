/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export interface AvailabilityType {
  name: string;
  color: string;
  mainVisId: string;
}
