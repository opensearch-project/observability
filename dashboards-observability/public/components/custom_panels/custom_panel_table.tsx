/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
/* eslint-disable react-hooks/exhaustive-deps */

import {
  EuiBreadcrumb,
  EuiButton,
  EuiContextMenuItem,
  EuiContextMenuPanel,
  EuiFieldSearch,
  EuiFlexGroup,
  EuiFlexItem,
  EuiHorizontalRule,
  EuiInMemoryTable,
  EuiLink,
  EuiOverlayMask,
  EuiPage,
  EuiPageBody,
  EuiPageContent,
  EuiPageContentHeader,
  EuiPageContentHeaderSection,
  EuiPageHeader,
  EuiPageHeaderSection,
  EuiPopover,
  EuiSpacer,
  EuiTableFieldDataColumnType,
  EuiText,
  EuiTitle,
} from '@elastic/eui';
import React, { ReactElement, useEffect, useState } from 'react';
import moment from 'moment';
import _ from 'lodash';
import { ChromeBreadcrumb } from '../../../../../src/core/public';
import {
  CREATE_PANEL_MESSAGE,
  CUSTOM_PANELS_DOCUMENTATION_URL,
} from '../../../common/constants/custom_panels';
import { UI_DATE_FORMAT } from '../../../common/constants/shared';
import { getCustomModal, DeletePanelModal } from './helpers/modal_containers';
import { CustomPanelListType } from '../../../common/types/custom_panels';
import { getSampleDataModal } from '../common/helpers/add_sample_modal';
import { pageStyles } from '../../../common/constants/shared';

/*
 * "CustomPanelTable" module, used to view all the saved panels
 *
 * Props taken in as params are:
 * loading: loader bool for the table
 * fetchCustomPanels: fetch panels function
 * customPanels: List of panels available
 * createCustomPanel: create panel function
 * setBreadcrumbs: setter for breadcrumbs on top panel
 * parentBreadcrumb: parent breadcrumb
 * renameCustomPanel: rename function for the panel
 * cloneCustomPanel: clone function for the panel
 * deleteCustomPanelList: delete function for the panels
 */

interface Props {
  loading: boolean;
  fetchCustomPanels: () => void;
  customPanels: CustomPanelListType[];
  createCustomPanel: (newCustomPanelName: string) => void;
  setBreadcrumbs: (newBreadcrumbs: ChromeBreadcrumb[]) => void;
  parentBreadcrumbs: EuiBreadcrumb[];
  renameCustomPanel: (newCustomPanelName: string, customPanelId: string) => void;
  cloneCustomPanel: (newCustomPanelName: string, customPanelId: string) => void;
  deleteCustomPanelList: (customPanelIdList: string[], toastMessage: string) => any;
  addSamplePanels: () => void;
}

export const CustomPanelTable = ({
  loading,
  fetchCustomPanels,
  customPanels,
  createCustomPanel,
  setBreadcrumbs,
  parentBreadcrumbs,
  renameCustomPanel,
  cloneCustomPanel,
  deleteCustomPanelList,
  addSamplePanels,
}: Props) => {
  const [isModalVisible, setIsModalVisible] = useState(false); // Modal Toggle
  const [modalLayout, setModalLayout] = useState(<EuiOverlayMask />); // Modal Layout
  const [isActionsPopoverOpen, setIsActionsPopoverOpen] = useState(false);
  const [selectedCustomPanels, setselectedCustomPanels] = useState<CustomPanelListType[]>([]);
  const [searchQuery, setSearchQuery] = useState('');

  useEffect(() => {
    setBreadcrumbs(parentBreadcrumbs);
    fetchCustomPanels();
  }, []);

  const closeModal = () => {
    setIsModalVisible(false);
  };

  const showModal = () => {
    setIsModalVisible(true);
  };

  const onCreate = async (newCustomPanelName: string) => {
    createCustomPanel(newCustomPanelName);
    closeModal();
  };

  const onRename = async (newCustomPanelName: string) => {
    renameCustomPanel(newCustomPanelName, selectedCustomPanels[0].id);
    closeModal();
  };

  const onClone = async (newName: string) => {
    cloneCustomPanel(newName, selectedCustomPanels[0].id);
    closeModal();
  };

  const onDelete = async () => {
    const toastMessage = `Custom Panels ${
      selectedCustomPanels.length > 1 ? 's' : ' ' + selectedCustomPanels[0].name
    } successfully deleted!`;
    const PanelList = selectedCustomPanels.map((panel) => panel.id);
    deleteCustomPanelList(PanelList, toastMessage);
    closeModal();
  };

  const createPanel = () => {
    setModalLayout(
      getCustomModal(
        onCreate,
        closeModal,
        'Name',
        'Create operational panel',
        'Cancel',
        'Create',
        undefined,
        CREATE_PANEL_MESSAGE
      )
    );
    showModal();
  };

  const renamePanel = () => {
    setModalLayout(
      getCustomModal(
        onRename,
        closeModal,
        'Name',
        'Rename Panel',
        'Cancel',
        'Rename',
        selectedCustomPanels[0].name,
        CREATE_PANEL_MESSAGE
      )
    );
    showModal();
  };

  const clonePanel = () => {
    setModalLayout(
      getCustomModal(
        onClone,
        closeModal,
        'Name',
        'Duplicate Panel',
        'Cancel',
        'Duplicate',
        selectedCustomPanels[0].name + ' (copy)',
        CREATE_PANEL_MESSAGE
      )
    );
    showModal();
  };

  const deletePanel = () => {
    const customPanelString = `operational panel${selectedCustomPanels.length > 1 ? 's' : ''}`;
    setModalLayout(
      <DeletePanelModal
        onConfirm={onDelete}
        onCancel={closeModal}
        title={`Delete ${selectedCustomPanels.length} ${customPanelString}`}
        message={`Are you sure you want to delete the selected ${selectedCustomPanels.length} ${customPanelString}?`}
      />
    );
    showModal();
  };

  const addSampledata = async () => {
    setModalLayout(
      getSampleDataModal(closeModal, async () => {
        closeModal();
        await addSamplePanels();
      })
    );
    showModal();
  };

  const popoverButton = (
    <EuiButton
      data-test-subj="operationalPanelsActionsButton"
      iconType="arrowDown"
      iconSide="right"
      onClick={() => setIsActionsPopoverOpen(!isActionsPopoverOpen)}
    >
      Actions
    </EuiButton>
  );

  const popoverItems: ReactElement[] = [
    <EuiContextMenuItem
      key="rename"
      disabled={customPanels.length === 0 || selectedCustomPanels.length !== 1}
      onClick={() => {
        setIsActionsPopoverOpen(false);
        renamePanel();
      }}
    >
      Rename
    </EuiContextMenuItem>,
    <EuiContextMenuItem
      key="duplicate"
      disabled={customPanels.length === 0 || selectedCustomPanels.length !== 1}
      onClick={() => {
        setIsActionsPopoverOpen(false);
        clonePanel();
      }}
    >
      Duplicate
    </EuiContextMenuItem>,
    <EuiContextMenuItem
      key="delete"
      data-test-subj="deleteContextMenuItem"
      disabled={customPanels.length === 0 || selectedCustomPanels.length === 0}
      onClick={() => {
        setIsActionsPopoverOpen(false);
        deletePanel();
      }}
    >
      Delete
    </EuiContextMenuItem>,
    <EuiContextMenuItem
      key="addSample"
      data-test-subj="addSampleContextMenuItem"
      onClick={() => {
        setIsActionsPopoverOpen(false);
        addSampledata();
      }}
    >
      Add samples
    </EuiContextMenuItem>,
  ];

  const tableColumns = [
    {
      field: 'name',
      name: 'Name',
      sortable: true,
      truncateText: true,
      render: (value, record) => (
        <EuiLink href={`${_.last(parentBreadcrumbs)!.href}${record.id}`}>
          {_.truncate(value, { length: 100 })}
        </EuiLink>
      ),
    },
    {
      field: 'dateModified',
      name: 'Last updated',
      sortable: true,
      render: (value) => moment(new Date(value)).format(UI_DATE_FORMAT),
    },
    {
      field: 'dateCreated',
      name: 'Created',
      sortable: true,
      render: (value) => moment(new Date(value)).format(UI_DATE_FORMAT),
    },
  ] as Array<EuiTableFieldDataColumnType<CustomPanelListType>>;

  return (
    <div style={pageStyles}>
      <EuiPage>
        <EuiPageBody component="div">
          <EuiPageHeader>
            <EuiPageHeaderSection>
              <EuiTitle size="l">
                <h1>Operational panels</h1>
              </EuiTitle>
            </EuiPageHeaderSection>
          </EuiPageHeader>
          <EuiPageContent id="customPanelArea">
            <EuiPageContentHeader>
              <EuiPageContentHeaderSection>
                <EuiTitle size="s">
                  <h3>
                    Panels
                    <span className="panel-header-count"> ({customPanels.length})</span>
                  </h3>
                </EuiTitle>
                <EuiSpacer size="s" />
                <EuiText size="s" color="subdued">
                  Use Operational panels to create and view different visualizations on ingested
                  observability data, using PPL (Piped Processing Language) queries.{' '}
                  <EuiLink external={true} href={CUSTOM_PANELS_DOCUMENTATION_URL} target="blank">
                    Learn more
                  </EuiLink>
                </EuiText>
              </EuiPageContentHeaderSection>
              <EuiPageContentHeaderSection>
                <EuiFlexGroup gutterSize="s">
                  <EuiFlexItem>
                    <EuiPopover
                      panelPaddingSize="none"
                      button={popoverButton}
                      isOpen={isActionsPopoverOpen}
                      closePopover={() => setIsActionsPopoverOpen(false)}
                    >
                      <EuiContextMenuPanel items={popoverItems} />
                    </EuiPopover>
                  </EuiFlexItem>
                  <EuiFlexItem>
                    <EuiButton
                      fill
                      onClick={() => createPanel()}
                      data-test-subj="customPanels__createNewPanels"
                    >
                      Create panel
                    </EuiButton>
                  </EuiFlexItem>
                </EuiFlexGroup>
              </EuiPageContentHeaderSection>
            </EuiPageContentHeader>
            <EuiHorizontalRule margin="m" />
            {customPanels.length > 0 ? (
              <>
                <EuiFieldSearch
                  fullWidth
                  data-test-subj="operationalPanelSearchBar"
                  placeholder="Search operational panel name"
                  value={searchQuery}
                  onChange={(e) => setSearchQuery(e.target.value)}
                />
                <EuiHorizontalRule margin="m" />
                <EuiInMemoryTable
                  loading={loading}
                  items={
                    searchQuery
                      ? customPanels.filter((customPanel) =>
                          customPanel.name.toLowerCase().includes(searchQuery.toLowerCase())
                        )
                      : customPanels
                  }
                  itemId="id"
                  columns={tableColumns}
                  tableLayout="auto"
                  pagination={{
                    initialPageSize: 10,
                    pageSizeOptions: [8, 10, 13],
                  }}
                  sorting={{
                    sort: {
                      field: 'dateModified',
                      direction: 'desc',
                    },
                  }}
                  allowNeutralSort={false}
                  isSelectable={true}
                  selection={{
                    onSelectionChange: (items) => setselectedCustomPanels(items),
                  }}
                />
              </>
            ) : (
              <>
                <EuiSpacer size="xxl" />
                <EuiText textAlign="center">
                  <h2>No Operational Panels</h2>
                  <EuiSpacer size="m" />
                  <EuiText color="subdued">
                    Use operational panels to dive deeper into observability
                    <br />
                    using PPL queries and insightful visualizations
                  </EuiText>
                </EuiText>
                <EuiSpacer size="m" />
                <EuiFlexGroup justifyContent="center">
                  <EuiFlexItem grow={false}>
                    <EuiButton
                      data-test-subj="customPanels__emptyCreateNewPanels"
                      fullWidth={false}
                      onClick={() => createPanel()}
                    >
                      Create panel
                    </EuiButton>
                  </EuiFlexItem>
                  <EuiFlexItem grow={false}>
                    <EuiButton fullWidth={false} onClick={() => addSampledata()}>
                      Add samples
                    </EuiButton>
                  </EuiFlexItem>
                </EuiFlexGroup>
                <EuiSpacer size="xxl" />
              </>
            )}
          </EuiPageContent>
        </EuiPageBody>
      </EuiPage>
      {isModalVisible && modalLayout}
    </div>
  );
};
