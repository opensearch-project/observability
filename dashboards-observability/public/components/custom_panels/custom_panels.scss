/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

 #panelView {
  .ppl-link {
    top: 118px;
    right: 644px;
  }
 }
 
 #panelViewInApp {
  .ppl-link {
    position: relative;
    top: -35px;
    left: 1386px;
  }
 }
 