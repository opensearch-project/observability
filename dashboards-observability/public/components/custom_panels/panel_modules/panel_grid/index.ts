/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export { PanelGrid } from "./panel_grid";
