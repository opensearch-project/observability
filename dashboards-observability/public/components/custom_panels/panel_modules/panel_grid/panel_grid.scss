/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

.react-grid-layout {
  & .react-grid-placeholder {
    background: $euiColorWarning;
  }
}

.full-width {
  min-width: 100%;
  max-width: 100%;
}
