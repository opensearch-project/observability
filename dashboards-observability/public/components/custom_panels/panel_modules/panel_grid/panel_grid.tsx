/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
/* eslint-disable react-hooks/exhaustive-deps */
/* eslint-disable no-console */

import _ from 'lodash';
import React, { useEffect, useState } from 'react';
import { Layout, Layouts, Responsive, WidthProvider } from 'react-grid-layout';
import useObservable from 'react-use/lib/useObservable';
import { CoreStart } from '../../../../../../../src/core/public';
import PPLService from '../../../../services/requests/ppl';
import { VisualizationContainer } from '../visualization_container';
import { VisualizationType } from '../../../../../common/types/custom_panels';
import { CUSTOM_PANELS_API_PREFIX } from '../../../../../common/constants/custom_panels';
import './panel_grid.scss';
import { mergeLayoutAndVisualizations } from '../../helpers/utils';

// HOC container to provide dynamic width for Grid layout
const ResponsiveGridLayout = WidthProvider(Responsive);

/*
 * PanelGrid - This module is places all visualizations in react-grid-layout
 *
 * Props taken in as params are:
 * http: http core service;
 * chrome: chrome core service;
 * panelId: OpenPanel Id
 * updateAvailabilityVizId: function to update application if availabilityViz is removed from panel
 * panelVisualizations: list of panel visualizations
 * setPanelVisualizations: function to set panel visualizations
 * editMode: boolean to check if the panel is in edit mode
 * pplService: ppl requestor service
 * startTime: start time in date filter
 * endTime: end time in date filter
 * onRefresh: boolean value to trigger refresh of visualizations
 * cloneVisualization: function to clone a visualization in panel
 * pplFilterValue: string with panel PPL filter value
 * showFlyout: function to show the flyout
 * editActionType: Type of action done while clicking the edit button
 */

interface PanelGridProps {
  http: CoreStart['http'];
  chrome: CoreStart['chrome'];
  panelId: string;
  updateAvailabilityVizId?: any;
  panelVisualizations: VisualizationType[];
  setPanelVisualizations: React.Dispatch<React.SetStateAction<VisualizationType[]>>;
  editMode: boolean;
  pplService: PPLService;
  startTime: string;
  endTime: string;
  onEditClick: (savedVisualizationId: string) => any;
  onRefresh: boolean;
  cloneVisualization: (visualzationTitle: string, savedVisualizationId: string) => void;
  pplFilterValue: string;
  showFlyout: (isReplacement?: boolean | undefined, replaceVizId?: string | undefined) => void;
  editActionType: string;
  setEditVizId?: any;
}

export const PanelGrid = (props: PanelGridProps) => {
  const {
    http,
    chrome,
    panelId,
    updateAvailabilityVizId,
    panelVisualizations,
    setPanelVisualizations,
    editMode,
    pplService,
    startTime,
    endTime,
    onEditClick,
    onRefresh,
    cloneVisualization,
    pplFilterValue,
    showFlyout,
    editActionType,
  } = props;
  const [currentLayout, setCurrentLayout] = useState<Layout[]>([]);
  const [postEditLayout, setPostEditLayout] = useState<Layout[]>([]);
  const [gridData, setGridData] = useState(panelVisualizations.map(() => <></>));
  const isLocked = useObservable(chrome.getIsNavDrawerLocked$());

  // Reset Size of Visualizations when layout is changed
  const layoutChanged = (currLayouts: Layout[], allLayouts: Layouts) => {
    window.dispatchEvent(new Event('resize'));
    setPostEditLayout(currLayouts);
  };

  const loadVizComponents = () => {
    const gridDataComps = panelVisualizations.map(
      (panelVisualization: VisualizationType, index) => (
        <VisualizationContainer
          key={panelVisualization.id}
          http={http}
          editMode={editMode}
          visualizationId={panelVisualization.id}
          savedVisualizationId={panelVisualization.savedVisualizationId}
          pplService={pplService}
          fromTime={startTime}
          toTime={endTime}
          onRefresh={onRefresh}
          onEditClick={onEditClick}
          cloneVisualization={cloneVisualization}
          pplFilterValue={pplFilterValue}
          showFlyout={showFlyout}
          removeVisualization={removeVisualization}
        />
      )
    );
    setGridData(gridDataComps);
  };

  // Reload the Layout
  const reloadLayout = () => {
    const tempLayout: Layout[] = panelVisualizations.map((panelVisualization) => {
      return {
        i: panelVisualization.id,
        x: panelVisualization.x,
        y: panelVisualization.y,
        w: panelVisualization.w,
        h: panelVisualization.h,
        static: !editMode,
      } as Layout;
    });
    setCurrentLayout(tempLayout);
  };

  // remove visualization from panel in edit mode
  const removeVisualization = (visualizationId: string) => {
    const newVisualizationList = _.reject(panelVisualizations, {
      id: visualizationId,
    });
    mergeLayoutAndVisualizations(postEditLayout, newVisualizationList, setPanelVisualizations);
  };

  // Save Visualization Layouts when not in edit mode anymore (after users saves the panel)
  const saveVisualizationLayouts = async (panelID: string, visualizationParams: any) => {
    return http
      .put(`${CUSTOM_PANELS_API_PREFIX}/visualizations/edit`, {
        body: JSON.stringify({
          panelId: panelID,
          visualizationParams,
        }),
      })
      .then(async (res) => {
        setPanelVisualizations(res.visualizations);
      })
      .catch((err) => {
        console.error(err);
      });
  };

  // Update layout whenever user edit gets completed
  useEffect(() => {
    if (editMode) {
      reloadLayout();
      loadVizComponents();
    }
  }, [editMode]);

  useEffect(() => {
    if (editActionType === 'save') {
      const visualizationParams = postEditLayout.map((layout) =>
        _.omit(layout, ['static', 'moved'])
      );
      saveVisualizationLayouts(panelId, visualizationParams);
      if (updateAvailabilityVizId) {
        updateAvailabilityVizId(panelVisualizations);
      }
    }
  }, [editActionType]);

  // Update layout whenever visualizations are updated
  useEffect(() => {
    reloadLayout();
    loadVizComponents();
  }, [panelVisualizations]);

  // Reset Size of Panel Grid when Nav Dock is Locked
  useEffect(() => {
    setTimeout(function () {
      window.dispatchEvent(new Event('resize'));
    }, 300);
  }, [isLocked]);

  useEffect(() => {
    loadVizComponents();
  }, [onRefresh]);

  useEffect(() => {
    loadVizComponents();
  }, []);

  return (
    <ResponsiveGridLayout
      layouts={{ lg: currentLayout, md: currentLayout, sm: currentLayout }}
      className="layout full-width"
      breakpoints={{ lg: 1200, md: 996, sm: 768, xs: 480, xxs: 0 }}
      cols={{ lg: 12, md: 12, sm: 12, xs: 1, xxs: 1 }}
      onLayoutChange={layoutChanged}
    >
      {panelVisualizations.map((panelVisualization: VisualizationType, index) => (
        <div key={panelVisualization.id}>{gridData[index]}</div>
      ))}
    </ResponsiveGridLayout>
  );
};
