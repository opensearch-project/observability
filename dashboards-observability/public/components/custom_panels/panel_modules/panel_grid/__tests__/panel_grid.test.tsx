/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import httpClientMock from '../../../../../../test/__mocks__/httpClientMock';
import { PanelGrid } from '../panel_grid';
import PPLService from '../../../../../services/requests/ppl';
import { VisualizationType } from '../../../../../../common/types/custom_panels';
import { coreStartMock } from '../../../../../../test/__mocks__/coreMocks';
import { waitFor } from '@testing-library/react';

describe('Panel Grid Component', () => {
  configure({ adapter: new Adapter() });

  it('renders panel grid component with empty visualizations', async () => {
    const http = httpClientMock;
    const core = coreStartMock;
    const panelId = '';
    const panelVisualizations: VisualizationType[] = [];
    const setPanelVisualizations = jest.fn();
    const editMode = false;
    const pplService = new PPLService(httpClientMock);
    const start = 'now-15m';
    const end = 'now';
    const onRefresh = false;
    const cloneVisualization = jest.fn();
    const pplFilterValue = '';
    const showFlyout = jest.fn();
    const editActionType = '';

    const wrapper = mount(
      <PanelGrid
        http={http}
        panelId={panelId}
        chrome={core.chrome}
        panelVisualizations={panelVisualizations}
        setPanelVisualizations={setPanelVisualizations}
        editMode={editMode}
        pplService={pplService}
        startTime={start}
        endTime={end}
        onRefresh={onRefresh}
        cloneVisualization={cloneVisualization}
        pplFilterValue={pplFilterValue}
        showFlyout={showFlyout}
        editActionType={editActionType}
      />
    );
    wrapper.update();

    await waitFor(() => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
