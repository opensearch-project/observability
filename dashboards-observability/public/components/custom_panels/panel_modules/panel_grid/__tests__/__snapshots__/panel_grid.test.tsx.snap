// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel Grid Component renders panel grid component with empty visualizations 1`] = `
<PanelGrid
  chrome={
    Object {
      "getIsNavDrawerLocked$": [MockFunction] {
        "calls": Array [
          Array [],
          Array [],
          Array [],
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": Observable {
              "_isScalar": false,
              "_subscribe": [Function],
            },
          },
          Object {
            "type": "return",
            "value": Observable {
              "_isScalar": false,
              "_subscribe": [Function],
            },
          },
          Object {
            "type": "return",
            "value": Observable {
              "_isScalar": false,
              "_subscribe": [Function],
            },
          },
          Object {
            "type": "return",
            "value": Observable {
              "_isScalar": false,
              "_subscribe": [Function],
            },
          },
        ],
      },
      "setBreadcrumbs": [MockFunction],
    }
  }
  cloneVisualization={[MockFunction]}
  editActionType=""
  editMode={false}
  endTime="now"
  http={[MockFunction]}
  onEditClick={[Function]}
  onRefresh={false}
  panelId=""
  panelVisualizations={Array []}
  pplFilterValue=""
  pplService={
    PPLService {
      "fetch": [Function],
      "http": [MockFunction],
    }
  }
  setPanelVisualizations={[MockFunction]}
  showFlyout={[MockFunction]}
  startTime="now-15m"
>
  <WidthProvider
    breakpoints={
      Object {
        "lg": 1200,
        "md": 996,
        "sm": 768,
        "xs": 480,
        "xxs": 0,
      }
    }
    className="layout full-width"
    cols={
      Object {
        "lg": 12,
        "md": 12,
        "sm": 12,
        "xs": 1,
        "xxs": 1,
      }
    }
    layouts={
      Object {
        "lg": Array [],
        "md": Array [],
        "sm": Array [],
      }
    }
    measureBeforeMount={false}
    onLayoutChange={[Function]}
  >
    <ResponsiveReactGridLayout
      breakpoints={
        Object {
          "lg": 1200,
          "md": 996,
          "sm": 768,
          "xs": 480,
          "xxs": 0,
        }
      }
      className="layout full-width"
      cols={
        Object {
          "lg": 12,
          "md": 12,
          "sm": 12,
          "xs": 1,
          "xxs": 1,
        }
      }
      layouts={
        Object {
          "lg": Array [],
          "md": Array [],
          "sm": Array [],
        }
      }
      onBreakpointChange={[Function]}
      onLayoutChange={[Function]}
      onWidthChange={[Function]}
      width={0}
    >
      <ReactGridLayout
        autoSize={true}
        className="layout full-width"
        cols={1}
        compactType="vertical"
        containerPadding={null}
        draggableCancel=""
        draggableHandle=""
        isDraggable={true}
        isResizable={true}
        layout={Array []}
        margin={
          Array [
            10,
            10,
          ]
        }
        maxRows={Infinity}
        onDrag={[Function]}
        onDragStart={[Function]}
        onDragStop={[Function]}
        onLayoutChange={[Function]}
        onResize={[Function]}
        onResizeStart={[Function]}
        onResizeStop={[Function]}
        preventCollision={false}
        rowHeight={150}
        style={Object {}}
        useCSSTransforms={true}
        verticalCompact={true}
        width={0}
      >
        <div
          className="react-grid-layout layout full-width"
          style={
            Object {
              "height": "10px",
            }
          }
        />
      </ReactGridLayout>
    </ResponsiveReactGridLayout>
  </WidthProvider>
</PanelGrid>
`;
